### DISCUSSION TITLE
Choices of twinBASIC

### DISCUSSION THREAD

**Discussion by ags1234** [twinbasic/twinbasic#1931] [General]
You have no idea how happy I was to see a modern IDE to continue where Visual Basic 6 left off!
And one under active development, yay!
Downloaded and saw another nice surprise: twinBASIC_IDE_BETA_622.zip is 24 MiB compressed and 76 MiB uncompressed.

But then... the bad news came.
1. The IDE is a Chrome Web app, not a real one. The feeling is different when working with Web apps.
2. The dependency required to run this 76 MiB Web app is Microsoft.WebView2.FixedVersionRuntime.130.0.2849.68.x64, which installed is 500+ MiB (!)
3. Not portable because of dependency.

If you really care about reviving vb6, which in itself is a good idea, it would have made much more sense to make it in vb6.

I mean the folks over in Pascal world wanted to continue where Delphi7 left off, so they made Lazarus, programmed in and compiled with free pascal. The sensible thing to do.

**Comment by WaynePhillipsEA** [comment]
The WebView2 dependency comes pre-packaged and complimentary with Windows 11. It's also bundled with Office, so really, most users already have it sitting there, twiddling its thumbs.

> it would have made much more sense to make it in vb6

Oh, absolutely... let's build something strictly 32-bit, forever tied to Windows. Cross-platform compatibility? Who needs it!

The IDE is an (offline) web app, and for a very good reason: future-proofing for cross-platform support. When the twinBASIC compiler grows up and learns to play nicely with other operating systems, with a nice and mature cross-platform UI/forms engine, maybe we’ll consider leaving the current setup. For now, this is what we’ve got, like it or lump it.

> I mean the folks over in Pascal world wanted to continue where Delphi 7 left off, so they made Lazarus, programmed in and compiled with free Pascal. The sensible thing to do.

Ah, sensibility - something we’re clearly not in the business of. Thanks for the tip.  :)

**Comment by WaynePhillipsEA** [comment]
Also, I will mention that at some point after the v1 release, we plan to open-source the IDE (note: not the compiler).  At that point, feel free to create a VB6 based compatible IDE based on it.

**Comment by wqweto** [comment]
> The IDE is a Chrome Web app, not a real one. The feeling is different when working with Web apps.

The feeling is exactly like coding in VS Code and most everyone likes the extra features this brings -- feartures one coming from VBx (or Notepad++) cannot even imagine and need some time to even discover and get used to.

> Not portable because of dependency.

Depenedency on Chrome based spin-off like Edge (i.e. webkit engine) is not a real dependency as every major OS has Chrome/chomium ported to it so setup will either not ship it or will ship a real world tested version for Windows/Linux/MacOS for free.

This is like complaining the IDE has dependency on KERNEL32.DLL or USER32.DLL -- a non-issue at present day and age.

**Comment by fafalone** [comment]
While in theory I'd also prefer a real desktop app because of the major issues with windowing even VSCode itself hasn't overcome, I think the choice at this stage was a no brainer: there's no ready-made desktop IDE kit that offers anywhere near the features of Monaco (AFAIK; I you know one, let us know), and I'm happy the language itself was the priority over adding another year or two to develop a full blown IDE or not offer a more modern IDE experience than basic VB6 level.

Remember too a lot of VB6 features depended on Internet Explorer. These are used rarely even compared to VB6 itself today, but I bet it's nonzero.

Portable generally means only system dependencies; as far as updated Windows 7, and Windows 10/11, it relies on a preinstalled system component so is portable. You can copy the tB folder to a new computer and just run, with only early unupdated 7 potentially not having WebView2.

Let's not let absolutely perfect be the enemy of freakin' fantastic 😄

**Comment by mansellan** [comment]
There's a couple of different factors at play here:

- Is your IDE native, or a website in disguise?
- Is the code editor in your IDE native, or a website in disguise?

Personally, even if the whole IDE is web-based, it's not a showstopper for me. I'm more concerned with how well it functions. I think I'm not alone, and that's why VS Code has done rather well.

That said, my preference is for the IDE "shell" to be native. It just gels better, and I've yet to see a truly outstanding window docking system in a web UI. But the IDE is really the least important priority for twinBASIC right now, and I don't want Wayne to spend too much time agonising over it. So long as it's good enough (it is), and better than the VB6 IDE (it is), it's fine for v1. Many people can write UI code. One person I know of can write a backwards-compatible VBx compiler from scratch. I'd much rather Wayne made a kick-ass compiler than a perfect IDE.

Nonetheless, the IDE is due to go open source in due course. When it does, people may well then take a shot at making native versions. But honestly, the one thing you don't want to try to replicate is the code editor. Monaco (the code editor in VSC) has taken the world by storm. It's utterly fantastic, you won't do better. By all means, wrap the other windows in native chrome, but it's insanity trying to write a better code editor from scratch.

**Comment by mansellan** [comment]
@WaynePhillipsEA 

> When the twinBASIC compiler grows up and learns to play nicely with other operating systems, with a nice and mature cross-platform UI/forms engine, maybe we’ll consider leaving the current setup.

Honestly, the only thing to consider is native UI wrappers for the IDE shell (and the various dockables). Why reinvent the code editor? Monaco is going to be a thing for a very, very long time, imho.

**Comment by ags1234** [comment]
It is not that I do not appreciate your work, please know that I do. Continuing from where vb6 left is a wonderful idea and I salute you for it. Breathing new life into an old language should be done with a very deep understanding of the past and a wide look into the future. What I am saying is the IDE could have used the knowledge those (probably) millions of vb6 programmers around the world have gathered in decades of writing and maintaining vb6. And the contributions to the code they could have done.

If your decisions are the right ones for this task, it means that when you'll open source the IDE you will find there are plenty of talented web developers interested in vb6 and a lot of experienced vb6 developers that know web technologies.

**Reply by wqweto** [reply]
There are no "millions of vb6 programmers around the world" anymore. There are zero VB6 developers capable of developing anything remotely similar to Monaco Editor.

There are zero VB6 developers capable of developing VB6 successor and that's why this project is not developed by VB6 developers.

**Reply by fafalone** [reply]
What kind of overly restrictive definition of 'VB6 developer' are you thinking of that wouldn't include Wayne as one lol

'Currently commercially or self-employed in a role where VB6 is the *primary* language' ? 

vbWatchDog worked with VB6; not sure you can make a tool that sophisticated work with something you don't know well enough to count as a developer in... so even before tB.

**Reply by wqweto** [reply]
Well, he chose C/C++ for the compiler for a reason I think.

**Reply by WaynePhillipsEA** [reply]
Multilingual :)

**Comment by rexxitall** [comment]
[Deleted expletive]

**Reply by WaynePhillipsEA** [reply]
Watch your language please.  If you disagree with something, by all means explain, but let's keep things friendly.
